testbench: dont build testbench, tools and topologies in the source directory

lgirdwood opened the pull request:
Currently all topologies, testbench and some tools are built within the FW source directory unlike FW which has separate binary destination directories outside of the main FW source directory.

This series moves all tools, testbench and topologies binaries to outside the source directory to the parent SOF "workspace" directory alongside the FW build output directories.

lgirdwood committed:
scripts: testbench: check for workspace environment variable

Allow both scripts to be executed from the SOF FW directory or the parent
SOF workspace directory using an environment variable or the default
SOF workspace.

This simplifies usage and opens the way for vscode integration.

Signed-off-by: Liam Girdwood <liam.r.girdwood@intel.com>

lgirdwood committed:
scripts: build tools: dont build in src dir

Dont built binaries in the source directory, build in the workspace.

Signed-off-by: Liam Girdwood <liam.r.girdwood@intel.com>

lgirdwood committed:
scripts: testbench_helper: fix indentation

Fix the indentation.

Signed-off-by: Liam Girdwood <liam.r.girdwood@intel.com>

lgirdwood committed:
scripts: testbench: move build output outside of source directory

Move all testbench build and test collateral outside of the SOF source
directory to the workspace directory.

Signed-off-by: Liam Girdwood <liam.r.girdwood@intel.com>

lgirdwood committed:
scripts: testbench: move test files outside of source directory

Move the test data files outside of the source directory and add
more context to their names for easier lookup. Do the same with
the outfile and generate it if we keep the temporary files.

Signed-off-by: Liam Girdwood <liam.r.girdwood@intel.com>

lgirdwood commented:
I expect this to break CI if we have hard coded assumptions of where to find certain binaries.
@singalsu pls take a look.

lyakh commented on scripts/sof-testbench-helper.sh:
```diff
@@ -177,16 +184,17 @@ else
     echo Converting raw output to "$OUTWAV"
     if [[ "$BITS" == "24" ]]; then
-	sox --encoding signed-integer -L -r "$RATE_OUT" -c "$CHANNELS_OUT" -b 32 "$OUTFILE1" "$OUTWAV" vol 256
+	    sox --encoding signed-integer -L -r "$RATE_OUT" -c "$CHANNELS_OUT" -b 32 \
```
The original indentation seems to have been correct. Also this and the following line shouldn't have equal indentation

lgirdwood replied:
This file is a mixture of tabs/spaces for indentation. Fixed this part.

lyakh commented on scripts/sof-testbench-build-profile.sh:
```diff
@@ -9,18 +9,28 @@ usage() {
-    echo "Error: environment variable SOF_WORKSPACE need to be set to top level sof directory"
+
+# check if we are in the sof FW directory
+if [ -f "Kconfig.sof" ]; then
```
can we maybe use something like `$(dirname "$0")` to identify the target directory, where we have to switch to? Then it'll be possible to call the script from anywhere

lgirdwood replied:
Ok, simplified this for non default installs - it will now check for ```SOF_WORKSPACE``` environment if set.

lyakh reviewed (commented):
In general a very good idea! Having tools' build directories withing the source tree was always quite inconvenient

kv2019i commented on scripts/rebuild-testbench.sh:
```diff
@@ -5,16 +5,27 @@
+# check if we are in the sof FW directory
+if [ -f "Kconfig.sof" ]; then
+    # now got to parent sof workspace directory
+    cd ..
```
Isn't this is a bit surprising to change the working directory when running the script? I don't see CWD restored at end. Wouldn't a less surprising method be to error if not run in correct directory?

PS and s/got to/go to/

lgirdwood replied:
Fixed the go to.
Expectation is that the script will be run from FW directory or the parent SOF workspace only. The directory change done by the script (in its own shell) will not change the users directory in their shell.

kv2019i reviewed (commented):
No blockers, @singalsu probably better to comment on the end-user impact on testbench scripting (see my comment).

This will need to be merged at the same time with Jenkins build changes. Another option is to make the build-outside-src an option, merge this in SOF and let Jenkins adopt this when it is ready (no flag day approach).

lgirdwood commented:
This depends on https://github.com/thesofproject/sof/pull/9870 so DNM as for now.

singalsu reviewed (commented):
Looks good so far, please add these changes to keep Matlab script tests working

```
diff --git a/tools/test/audio/comp_run.sh b/tools/test/audio/comp_run.sh
index 096ff2c34..6beaf74f5 100755
--- a/tools/test/audio/comp_run.sh
+++ b/tools/test/audio/comp_run.sh
@@ -132,13 +132,13 @@ run_testbench ()
 parse_args "$@"
 
 # Path to topologies
-TPLG_DIR=../../build_tools/test/topology
+TPLG_DIR=../../../../build_tools/test/topology
 
 # Testbench path and executable
 if [[ -z $XTRUN ]]; then
-    PATH_TESTBENCH=../../testbench/build_testbench/install/bin/$TESTBENCH
+    PATH_TESTBENCH=../../../../build_testbench/install/bin/$TESTBENCH
 else
-    BUILD_DIR=../../testbench/build_xt_testbench
+    BUILD_DIR=../../../../build_xt_testbench
     PATH_TESTBENCH="$BUILD_DIR"/$TESTBENCH
     source "$BUILD_DIR"/xtrun_env.sh
     XTRUN_CMD=$XTENSA_PATH/$XTRUN
@@ -168,7 +168,7 @@ PIPELINES=
     [[ $DIRECTION == "playback" ]] && PIPELINES="-p 1,2"
     [[ $DIRECTION == "capture" ]] && PIPELINES="-p 3,4"
     TPLGFN=sof-hda-benchmark-${COMP}${BITS_IN}.tplg
-    TPLG_DIR="../../build_tools/topology/topology2/development"
+    TPLG_DIR="../../../../build_tools/topology/topology2/development"
     TPLG_BUILD_TIP="Please run scripts/build-tools.sh"
 }

```

lgirdwood commented:
Changes for V2:
1) Check for ```SOF_WORKSPACE``` environment for non default installs.
2) Added @singalsu patch.

singalsu reviewed (commented):
Good cleanup, thanks!

lrgirdwo committed:
build-tools: deploy: Add deployment option for topology2

Add a simple deployment option fro topology2 to copy topologies to
the staging build directory. Simple today as only Intel use tplg2.

Signed-off-by: Liam Girdwood <liam.r.girdwood@linux.intel.com>

lgirdwood commented:
depends on https://github.com/thesofproject/sof/pull/9870

lgirdwood commented:
Rebased now that #9870 is merged

kv2019i reviewed (commented):
Needs Jenkins update to merge.

lgirdwood commented:
> Needs Jenkins update to merge.

Yeah - hopefully the last one. @cgturner1 pls see jenkins build i.e. we no longer build tools in the src directory.
```
/usr/bin/mkdir -p -m 0755 /home/cwd_user/work/sof/tools/var/lib/alsa
make[3]: Leaving directory '/home/cwd_user/work/sof/alsa-utils'
make[2]: Leaving directory '/home/cwd_user/work/sof/alsa-utils'
make[1]: Leaving directory '/home/cwd_user/work/sof/alsa-utils'
Build and installation complete for /home/cwd_user/work/sof/alsa-utils
All repositories processed.
mkdir: cannot create directory '/home/sof/work/sof.git/../build-tools-tplg': Permission denied
```

cgturner1 commented:
ok will need to take a look...a lot of the issues stem from the user settings (in scripts/sudo-cwd.sh) being done in the docker which seem to be from the way it mounts uses the local volumes as a workspace

and it sems to restrict the user "cwd" from creating directories in certain places ...ideally this docker should run as the sof user and we would be fine..i'll need to figure out the change to get this to cooperate at least to get the mkdir to work.

will run some tests and update when i get this updated

lyakh commented on scripts/sof-testbench-build-profile.sh:
```diff
@@ -9,18 +9,32 @@ usage() {
+    # Environment variable exists and has a value
+    BASE_DIR="$SOF_WORKSPACE"
+fi
+cd "$BASE_DIR"
```
cannot we use something like
```
SCRIPTS_DIR=$(dirname "${BASH_SOURCE[0]}")
SOF_DIR=$SCRIPTS_DIR/../
```
like other SOF scripts do to avoid guessing?

lgirdwood replied:
So we have a bit of complexity today as CI has some hard coded rules (with config outside this repo). Long term trying to move to an environment variable where the scripts could be executed in any directory under ```SOF_WORKSPACE```.
I do like this idea, will make the change.

intel-sofci commented:
<!-- summary-comment:test-summary -->
### PR 9871: test results

Run date: 2026-08-31 16:18 UTC

Tested commit: [43838ef81671fbb32c42b4b6c9e8b8e90eb61347](https://github.com/thesofproject/sof/commit/43838ef81671fbb32c42b4b6c9e8b8e90eb61347)

![mtl pass rate](https://img.shields.io/badge/mtl-critical_failure-black) ![lnl pass rate](https://img.shields.io/badge/lnl-passed-brightgreen) ![ptl pass rate](https://img.shields.io/badge/ptl-passed-brightgreen) ![wcl pass rate](https://img.shields.io/badge/wcl-passed-brightgreen) ![nvl pass rate](https://img.shields.io/badge/nvl-critical_failure-black)
